File: haskell/protein-translation/src/ProteinTranslation.hs
module ProteinTranslation(proteins) where

import Data.List.Split (chunksOf)

proteins :: String -> Maybe [String]
proteins = fmap (takeWhile (/="STOP")) . traverse proteinLookup . chunksOf 3

proteinLookup :: String -> Maybe String
proteinLookup p
    | p == "AUG"                            = Just "Methionine"
    | p `elem` ["UUU", "UUC"]               = Just "Phenylalanine"
    | p `elem` ["UUA", "UUG"]               = Just "Leucine"
    | p `elem` ["UCU", "UCC", "UCA", "UCG"] = Just "Serine"
    | p `elem` ["UAU", "UAC"]               = Just "Tyrosine"
    | p `elem` ["UGU", "UGC"]               = Just "Cysteine"
    | p == "UGG"                            = Just "Tryptophan"
    | p `elem` ["UAA", "UAG", "UGA"]        = Just "STOP"
    | otherwise                             = Nothing

File: cpp/protein-translation/protein_translation.cpp
#include "protein_translation.h"
#include <string>
#include <unordered_map>
#include <vector>

std::vector<std::string> protein_translation::proteins(const std::string &s) {
    const static std::unordered_map<std::string, std::string> codon_to_protein =
        {{"AUG", "Methionine"},    {"UUU", "Phenylalanine"},
         {"UUC", "Phenylalanine"}, {"UUA", "Leucine"},
         {"UUG", "Leucine"},       {"UCU", "Serine"},
         {"UCC", "Serine"},        {"UCA", "Serine"},
         {"UCG", "Serine"},        {"UAU", "Tyrosine"},
         {"UAC", "Tyrosine"},      {"UGU", "Cysteine"},
         {"UGC", "Cysteine"},      {"UGG", "Tryptophan"},
         {"UAA", "STOP"},          {"UAG", "STOP"},
         {"UGA", "STOP"}};

    std::vector<std::string> v;
    for (size_t i = 0; i < s.length(); i += 3) {
        const auto &pos = codon_to_protein.find(s.substr(i, 3));
        if (pos != codon_to_protein.end()) {
            if (pos->second == "STOP") {
                break;
            }

            v.emplace_back(pos->second);
        }
    }

    return v;
}
